design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,172 @@
+# Synchronize a Set of Labels on the BaremetalHost with the corresponding Kubernetes Node
+
+There are use cases where certain information is inherently tied to the BMH but at the same time, is valuable to the users/operators/schedulers in the K workload cluster. For example, a user may wish to place her workloads across hosts that are in different failure zones. In Kubernetes, labels on Node objects are the primary means by which to solve this problem. The proposal is for CAPM3 to synchronize a specific set of labels placed on a BMH with labels on the corresponding Kubernetes Node running on that BMH.
+
+
+# User Stories
+
+As a user I want replicas of a specific workload spread out across different server racks or geographical locations.
+
+As a user, I would like to place my security sensitive workloads on machines that meet specific security requirements. For example, certain hosts may be in a fortified zone within a data center or certain hosts may have strong hardware based security mechanisms in place (e.g. hardware attestation via TPM).
```

Comment:
It seems like both of these use cases could be achieved by using different MachineDeployments, based on physical location or security capabilities. Have you explored that as an option?

In our use case, we wanted to add labels based on data exposed through inspection on the host. That same information may also be available through node-feature-discovery, so we are looking at using that tool instead.

Reply:
I think both approaches you mentioned are complementary to what is being proposed here. The differences I see are:

1. MD labels are only applicable at creation time. Adding/Removing a new label requires a complete MD rollout. This is highly undesirable in a baremetal environment. The proposed approach would offer more flexibility. 
2. MD grouping may not always provide the needed granularity; effectively MD == workload type in what you suggested. Depending on the how granular the user wants their labeling, this may lead to a large number of MDs. 
3. There are cases where a logical grouping like MD is not appropriate when trying to capture physical grouping. For example, you may want a specific label on all hosts in the same rack but logically hosts in the rack may belong to different MDs. 
4. Certain information may not be discoverable by host inspection alone. For example, user intent cases such as I want to label certain hosts as temporary so only workloads that can tolerate interruptions are scheduled there. Also, the label sync may be a convenient alternative for users who don't want to implement/deploy controllers for feature discovery. 
5. There are security concerns with kubelet's `--node-labels` flag (as mentioned in this proposal). This is not a deal breaker but definitely undesirable. 

The CAPI proposal mentioned in related section would allow for labels to be added/removed without a rollout, but that is still in the proposal phase.

I can add some (or all) of the points above to the proposal if they provide more clarity.

Reply:
It does sound like MachineDeployments will not achieve the goals. Listing that in the alternatives section would be good.

Reply:
Done.